Add tests for ConsoleReporter.

piyush-jaiswal opened the pull request:
This commit adds tests for the ConsoleReporter which includes tests for the feature added by [#5667](https://github.com/scala/scala/pull/5667).

piyush-jaiswal commented:
Added 2 files, am not certain about the coding style.

som-snytt commented:
I think it's the git style. You didn't intend to commit all the extra files.

piyush-jaiswal commented:
Yeah, exactly. Anyways, just to clarify, I have signed the CLA.

piyush-jaiswal commented:
@adriaanm, could you review it?

som-snytt commented:
Since `ConsoleReporter` takes a reader and writer, the unit test should supply them and not use `Console`.

(Using `Console` in particular reduces test isolation, a problem if you want to run many tests without forking.)

This test could also subsume the new test at https://github.com/scala/scala/pull/5667 where probably an end-to-end test is overkill.

Note that maxerr isn't fixed at 100 now, so setting it explicitly for the test is useful (instead of testing the default).

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,171 @@
+
+      // finish test
+      reporter.ERROR.count = 0
+      reporter finish;    // don't know why but I'm requiring ';' here, otherwise the file does not compile
```
I'd suggest not enabling postfixOps at all.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,171 @@
+      val reporter = new ConsoleReporter(new Settings())
+
+      // printMessage test
+      reporter printMessage("Hello World!")
```
I'd suggest using dot-selection, especially with parens.
```
reporter.printMessage("hello, world.")
reporter printMessage "hello, world."    // Recently, I use this style less, but community hasn't outlawed it.
```

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,171 @@
+      err_out.reset
+
+      reporter.WARNING.count = 3
+      reporter.ERROR.count = 10
```
Probably when the reporter API is nailed down some more, this should be disallowed. Maybe Reporter should have `protected increment(severity)`.

piyush-jaiswal commented:
Pardon me @som-snytt , I didn't know you were my assigned reviewer. I will make the changes as suggested but in the meanwhile, my midterms are around the corner. Is it okay if this can wait for some days?

som-snytt commented:
Absolutely, reviewing and updating are on a volunteer basis!

piyush-jaiswal commented:
Hey @som-snytt, I made the changes. Could you review and suggest further changes if needed.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,185 @@
+  @Test
+  def unitTest: Unit = {
+    val noPos = NoPosition
+    val path = "test/junit/scala/tools/nsc/reporters/file_ConsoleReporterTest.txt"
```
I'm not clear on why a file is needed. Source can be just a byte array.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,185 @@
+    // printMessage test
+    reporter.printMessage("Hello World!")
+    assertEquals("Hello World!\n", err_out.toString)
+    err_out.reset
```
Could refactor this pattern to a method.

piyush-jaiswal replied:
I'm not sure that I get you correctly. Do you want me to group the related tests inside a method, for example the tests for printMessage inside a single method?

piyush-jaiswal commented:
Made the changes. Still not sure, if it's exactly what you wanted.

piyush-jaiswal commented:
Hey @som-snytt,  could you review the changes?

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,188 @@
+    val input = new StringReader("s\n")
+    val input_reader = new BufferedReader(input)
+    val setting = new Settings()
+    setting.maxerrs.tryToSet(List("50"))  //changing maxerrs
```
`settings.maxerrs.value = 50`

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,188 @@
+    def printMessageTest(): Unit = {
+      reporter.printMessage("Hello World!")
+      assertEquals("Hello World!\n", err_out.toString)
+      err_out.reset
```
You could capture this repeated pattern in a routine that performs the assert and the reset. I often see simple helpers like that in unit tests.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,188 @@
+
+      // printColumnMarkerTest test with DefinedPosition
+      reporter.printColumnMarker(posWithSource)
+      assertEquals("    ^\n", err_out.toString)
```
This assumes the platform line ending, perhaps. `text.lines` converts to lines.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,188 @@
+    @Test
+    def displayPromptTest(): Unit = {
+      val stack = new ByteArrayOutputStream
+      System.setErr(new PrintStream(stack))  // used for checking the stack trace of displayPrompt()
```
`Console.withErr`. Again, it's easier to encapsulate with a method that does the setup. Allocating a BAOS for every test is easier than `stack.reset`. As a style matter, maybe the idea is that it's worth reusing a heavyweight object?

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,188 @@
+        val input = new StringReader("r\n")
+        val input_reader = new BufferedReader(input)
+        val reporter = new ConsoleReporter(new Settings(), input_reader, output_writer)
+        reporter.displayPrompt
```
Not a huge deal, but I'd expect parens to demonstrate a side-effecting application. That's the standard advice. This looks like it just accesses a value.

som-snytt reviewed (commented):
I made a couple of small comments. I'm trying out the "review" button for the first time, which really complicates the workflow. (note to github)

You need to rebase and clean up the git history, which includes spurious merge commits. They want to see your one squashed commit before they merge. (as a rule)

piyush-jaiswal commented:
/rebuild

piyush-jaiswal commented:
/rebuild

som-snytt commented:
You have to fix the git history. Your PR should show your one commit.

piyush-jaiswal commented:
I didn't push form the '2.12.x' branch after rebasing it, think that was the root of the problem. Thanks @som-snytt !
I added the protected increment and reset severity functions in ConsoleReporter and also modified the tests. Hope you can take a look at it.

adriaanm commented:
@piyush-jaiswal, note that you seem to have accidentally committed a bunch of unrelated files. Please also review our [guidelines](https://github.com/scala/scala/blob/2.12.x/CONTRIBUTING.md#clean-commits-clean-history) on how to format a good commit message. As it stands, your commit title is too long, is not in the present tense, and there's no elaboration in the commit message itself (for example: what does this do? why? what was wrong before?)

Thanks for reviewing @som-snytt! When you feel this PR is ready, feel free to merge.

piyush-jaiswal committed:
Add tests for ConsoleReporter.

piyush-jaiswal commented:
Made the changes. Kindly point out other issues if present.

piyush-jaiswal commented:
/rebuild

piyush-jaiswal commented:
Hey @som-snytt , how are you? I've made some changes since we last communicated. In particular I've added the increment and reset severity functions in the ConsoleReporter as you had earlier suggested.

 I still couldn't figure out how to use `Console.withErr` instead of `System.setErr` for the `displayPromptTest` as I am checking for the stack trace which I think is printed in the standard error stream System.err. Could you go ahead and have a look?

som-snytt commented:
Sorry, just a busy time, albeit not midterms. If I suggested adding API to ConsoleReporter just for a test, don't listen to that guy. Also, there was just a commit to let it take two writers, for error and echo. That might affect your test.

You can actually fix something by changing it to use `printStackTrace(writer)`. It becomes testable and correct. I guess you can't test the abort option.

piyush-jaiswal commented:
Okay, I've changed my tests according to the commit and yeah I don't think it's possible to test the abort option too.

piyush-jaiswal commented:
review @som-snytt ?

piyush-jaiswal commented:
Is there any further issue present? I am finding it difficult to spotting them on my own.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,211 @@
+      if(echoWriterOut == null)
+        new ConsoleReporter(new Settings(), reader, writer) 
+      else
+      {
```
Repo style is space after if `if (` and open brace on same line `else {`.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,211 @@
+
+
+  def testHelper(checkBlank: Boolean, hasPos: Boolean, message: String, severity: String): Unit = {
+    if(checkBlank) assertEquals("", writerOut.toString)
```
the flag is redundant with `message = ""`. Possibly `message.isEmpty && severity.isEmpty`.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,211 @@
+      }
+    }
+    writerOut.reset
+  }
```
The test rig could be: `testHelper(pos: Position = NoPosition, msg: String, severity: String = "")(test: Position => Unit)`.

`testHelper(msg = "hello, world")(_ => reporter.printMessage("hello, world"))`.

`testHelper(myPos, "test")(reporter.print(_, "test", INFO))` or similar.

The `hasPos` flag is `pos.isDefined`.

I lose track of the `true, false` pairs in the tests.

piyush-jaiswal replied:
I'm not sure if we require `Position` as an argument of the function stored in the second parameter, `test` of `testHelper`.

We could just do:
`def testHelper(pos: Position = NoPosition, msg: String, severity: String = "")(test: () => Unit)`

som-snytt replied:
I think I was trying to avoid repeating the pos info. My second example reads well; you'd only need arrow syntax to ignore the pos in the test function.

But yes, you could just:

`testHelper(myPos, "test")(reporter.print(myPos, "test", INFO))`

with by-name param instead of a function:

`def testHelper(...)(test: => Unit)`

Either way, I think it's worth the trouble, because they want to do more tricks with reporters, like filtering, so it's likely your test will get exercised and expanded.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,211 @@
+    testHelper(false, false, "test", "")
+
+    reporter.print(noPos, "test", reporter.WARNING)
+    testHelper(false, false, "test", "warning: ")
```
The `false, false`, which I've forgotten what it means, is noise.

Sometimes a little table of truth values in code is tidy, but not as args. It can help to name them, `f(hasPos = false, isBlank = false, x, y)`, and in non-test code the repo requires named boolean args for this reason.

Eliminating the noise would take this to another level.

som-snytt commented on src/compiler/scala/tools/nsc/reporters/ConsoleReporter.scala:
```diff
@@ -91,4 +91,4 @@ class ConsoleReporter(val settings: Settings, reader: BufferedReader, writer: Pr
 
   override def finish() = printSummary()
-}
+}
```
Probably restore terminal NL.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,211 @@
+
+    reporter.WARNING.count = 26
+    reporter.display(posWithSource, "Testing display for maxwarns to fail", reporter.WARNING)
+    testHelper(true, false, "", "")
```
The one thing tested by `test/files/neg/maxerrs.scala` is that output is suppressed but the error or warning over the max to display is still counted, and not simply ignored. Here, the count is bumped.

piyush-jaiswal commented:
I've tried to make the changes. Please review. Thanks!

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,163 @@
+      Option(echoWriterOut) match {
+        case None => new ConsoleReporter(new Settings(), reader, writer)
+        case Some(_) => new ConsoleReporter(new Settings(), reader, writer, new PrintWriter(echoWriterOut))
+      }
```
This is an idiom you'll see nowhere. I won't nitpick default to null, but you can just `case null =>` or just `val echo = if (out != null) new PrintWriter(out) else writer`. You might as well name the params `errOut` and `echoOut` since they aren't Writers.

piyush-jaiswal replied:
I think I read somewhere that `null` was provided in scala for java interop and that `Option` is better to use in scala. As it may be, I think I'll use `null =>` as it utilizes one of the constructors in ConsoleReporter.

som-snytt replied:
To avoid both `null` and boxing, method overload: `f(out)` calls `f(out, out)`. Sometimes the default arg can be an empty value of the desired type.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,163 @@
+    testHelper(msg = "test")(_ => reporter.print(NoPosition, "test", reporter.INFO))
+    testHelper(msg = "test", severity = "warning: ")(_ => reporter.print(NoPosition, "test", reporter.WARNING))
+    testHelper(msg = "test", severity = "error: ")(_ => reporter.print(NoPosition, "test", reporter.ERROR))
+    testHelper(posWithSource, msg = "test")(_ => reporter.print(posWithSource, "test", reporter.INFO))
```
Since your helper takes `Pos => Unit`, this would be
```
testHelper(posWithSource, msg = "test")(reporter.print(_, "test", reporter.INFO))
```
If you find that finicky, then helper could just take `(body: => Unit)`, and maybe that's easier to read, too:
```
testHelper(posWithSource, msg = "test")(reporter.print(posWithSource, "test", reporter.INFO))
```

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,163 @@
+    reporter.ERROR.count += 1
+    testHelper(posWithSource, msg = "Testing display for maxerrs to pass", severity = "error: ")(_ => reporter.display(posWithSource, "Testing display for maxerrs to pass", reporter.ERROR))
+    reporter.ERROR.count += 1
+    testHelper(msg = "")(_ => reporter.display(posWithSource, "Testing display for maxerrs to fail", reporter.ERROR))
```
Since you're testing `display` (and not `reporter.error`), I'd set `maxerrs = 1` and `ERROR.count = 0` explicitly in the test so we can see the conditions of the test. I don't know what the current count is, does it depend on order tests are run? etc.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,163 @@
+    assertEquals("three warnings found", it.next)
+    assertEquals("10 errors found", it.next)
+    writerOut.reset
+  }
```
The tests assume test isolation. Probably I wouldn't have used a shared reporter, since it's cheap to construct. Each test could use a throw-away reporter. Here, for instance, the counts are bumped up and other tests could see that, depending on how the test is run (one Test instance per test method or one instance for all, test order, etc).

piyush-jaiswal replied:
Should I use a separate `BAOS` for each test as well and pass it as an argument in `testHelper` or let it remain a global as it is now?

som-snytt commented:
Just a couple of small comments to push it across. It's rude at this point to ask not to use a shared reporter across test methods, so I'll leave that to you, so long as the tests actually pass. For future reference, I avoid sharing the "system under test" unless it's expensive to construct.

adriaanm commented:
Good work, @piyush-jaiswal -- thanks. Thank you for the in-depth reviewing, @som-snytt!

som-snytt commented:
I'm pretty sure this PR hit all of the hardest problems in computer science, or I might be off by one.

This test will be so useful when someone adds reporter features. :+1:

piyush-jaiswal commented:
Updated. I've used the same `BAOS` for constructing all the separate reporters ATM.

som-snytt commented on test/junit/scala/tools/nsc/reporters/ConsoleReporterTest.scala:
```diff
@@ -0,0 +1,173 @@
+        assertEquals("    ^", it.next)
+      }
+    }
+    writerOut.reset
```
Technically, we should have put the reset in a finally block. If it fails, it throws, so the next test to run could see the dirty buffer. Probably the test class is reinstantiated; I think junit does that.

som-snytt commented:
LGTM

piyush-jaiswal commented:
Thanks @adriaanm and thanks @som-snytt  for taking the time to review this PR. This definitely was a great learning experience for me!
